tests/base.py

```diff
@@ -51,3 +51,114 @@ def create_fake_config(name, **kwargs):
+
+        out = {}
+        for key in values.keys():
+            if key in including or not including:
+                out[key] = values[key]
+            if key in out and (key in excluding or (including and key not in including)):
+                del out[key]
+        return out
+
+    def test_valid_config(self):
+        """
+        This is essentially a smoke test to ensure the following:
+        1) An instance of the correct class is created from the valid configuration info
+        2) That the configuration is in the proper state pre and post validation
+        """
+        config = self.CONFIG_CLASS.from_dict(self.VALID_CONFIG, "test", None)
+
+        self.assertIsInstance(config, self.CONFIG_CLASS, 'Wrong type of config, was "type" set correctly in "VALID_CONFIG"')
+
+        # Having been updated, the config section needs validation
+        self.assertEqual(config.state, ValidationState.NEEDS_VALIDATION)
+
+        messages = config.validate()
+
+        # The config state should be valid, post validation with valid config information
+        self.assertEqual(config.state, ValidationState.VALID)
+
+        # There should be no error messages if this config section is valid
+        self.assertFalse(any(message[0] == 'error' for message in messages))
+
+    def test_missing_required_no_default(self):
+        """
+        This tests that when a required key, which has no default, is missing, the config section
+        becomes invalid post-validation.
+        """
+        config = self.CONFIG_CLASS("test", None)
+        # A sorted list of keys that are required, but have no default
+        target_keys = sorted(list(config._required_keys - set(config.defaults.keys())))
+        config_values = self._modified_dict(self.VALID_CONFIG, excluding=target_keys)
+        config.update(**config_values)
+
+        messages = config.validate()
+
+        self.assertEqual(config.state, ValidationState.INVALID)
+        self.assertTrue(any(message[0] == 'error' for message in messages))
+
+    def test_missing_required_with_default(self):
+        """
+        Test that a config section given a set of values which are valid other than missing required
+        keys which have defaults, will be valid post validation, and that the values for the
+        defaults are used.
+
+        NOTE: This does not account for modifications of default values made during validation.
+        """
+        config = self.CONFIG_CLASS("test", None)
+        target_keys = sorted(list(config._required_keys & set(config.defaults.keys())))
+        config_values = self._modified_dict(self.VALID_CONFIG, excluding=target_keys)
+        config.update(**config_values)
+
+        config.validate()
+
+        self.assertEqual(config.state, ValidationState.VALID)
+
+        for key in target_keys:
+            self.assertIn(key, config, "Key '%s' is missing from config")
+            self.assertEqual(config[key], config.defaults[key])
```

Comment:
I would agree but the defaults, as well as the required values are determined entirely by calls the add_key for the config class.

Do you think it would be enough to verify that the add_key method does what we expect? (and that the validation state of a config class given those calls to add_key does what we expect)?

Reply:
What about adding here something like this:

```python
self.assertEqual(config[key], self.DEFAULT_VALUES[key])
```

And subclass would contain something like this:

```python
DEFAULT_VALUES = { 'is_hypervisor': True }
```